src/web/demo.rkt

```diff
@@ -390,155 +517,101 @@
-      ;; Is this ok because we are multithreaded now?
-      (set-seed! seed)
-      (semaphore-wait (run-sample hash formula))
+      (define hash (sha1 (open-input-string formula-str)))
+      (semaphore-wait (run-sample hash formula seed))
       (hash-ref *completed-jobs* hash))))
 
+(define (run-analyze hash formula seed pcontext sample)
+  (hash-set! *jobs* hash (*timeline*))
+  (define sema (make-semaphore))
+  (thread-send *worker-thread* (list 'errors hash formula sema seed pcontext sample))
+  sema)
+
 (define analyze-endpoint
   (post-with-json-response
     (lambda (post-data)
-      (define formula (read-syntax 'web (open-input-string (hash-ref post-data 'formula))))
+      (define formula-str (hash-ref post-data 'formula))
+      (define formula (read-syntax 'web (open-input-string formula-str)))
       (define sample (hash-ref post-data 'sample))
       (define seed (hash-ref post-data 'seed #f))
-      (eprintf "Analyze job started on ~a..." formula)
-
       (define test (parse-test formula))
       (define pcontext (json->pcontext sample (test-context test)))
-      (define result (run-herbie 'errors test #:seed seed #:pcontext pcontext
-                                 #:profile? #f #:timeline-disabled? #t))
-      (define errs (job-result-backend result))
+      (define hash (sha1 (open-input-string formula-str)))
+      (semaphore-wait (run-analyze hash formula seed pcontext sample))
+      (hash-ref *completed-jobs* hash))))
 
-      (eprintf " complete\n")
-      (hasheq 'points errs))))
+(define (run-exacts hash formula seed sample)
+  (hash-set! *jobs* hash (*timeline*))
+  (define sema (make-semaphore))
+  (thread-send *worker-thread* (list 'exacts hash formula sema seed sample))
+  sema)
 
 ;; (await fetch('/api/exacts', {method: 'POST', body: JSON.stringify({formula: "(FPCore (x) (- (sqrt (+ x 1))))", points: [[1, 1]]})})).json()
-(define exacts-endpoint 
+(define exacts-endpoint
   (post-with-json-response
     (lambda (post-data)
-      (define formula (read-syntax 'web (open-input-string (hash-ref post-data 'formula))))
+      (define formula-str (hash-ref post-data 'formula))
+      (define formula (read-syntax 'web (open-input-string formula-str)))
       (define sample (hash-ref post-data 'sample))
       (define seed (hash-ref post-data 'seed #f))
-      (eprintf "Ground truth job started on ~a..." formula)
-
-      (define test (parse-test formula))
-      (define pcontext (json->pcontext sample (test-context test)))
-      (define result (run-herbie 'exacts test #:seed seed #:pcontext pcontext
-                                 #:profile? #f #:timeline-disabled? #t))
-      (define exacts (job-result-backend result))
+      (define hash (sha1 (open-input-string formula-str)))
+      (semaphore-wait (run-exacts hash formula seed sample))
+      (hash-ref *completed-jobs* hash))))
 
-      (eprintf " complete\n")
-      (hasheq 'points exacts))))
+(define (run-evaluate hash formula seed sample)
```

Comment:
I do not think we need all these `run-xxx` methods, seeing as they are all basically the same, but I think the plan in this PR is to make them uniform so that later we can deduplicate.

Reply:
Will do in next PR